### ISSUE TITLE
`trainer.predict()` fails to find the last checkpoint from `ModelCheckpoint` callback

### ISSUE THREAD

**Issue by olipinski** [Lightning-AI/lightning#17606]
### Bug description

When using `ModelCheckpoint` callback, and then trying to use `trainer.predict()` no checkpoints are found.

### What version are you seeing the problem on?

v2.0

### How to reproduce the bug

```python
import os
from pytorch_lightning.callbacks import ModelCheckpoint
from torch import optim, nn, utils
from torchvision.datasets import MNIST
from torchvision.transforms import ToTensor
import pytorch_lightning as pl

# define any number of nn.Modules (or use your current ones)
encoder = nn.Sequential(nn.Linear(28 * 28, 64), nn.ReLU(), nn.Linear(64, 3))
decoder = nn.Sequential(nn.Linear(3, 64), nn.ReLU(), nn.Linear(64, 28 * 28))

# define the LightningModule
class LitAutoEncoder(pl.LightningModule):
    def __init__(self, encoder, decoder):
        super().__init__()
        self.encoder = encoder
        self.decoder = decoder

    def forward(self, batch):
        x, y = batch
        x = x.view(x.size(0), -1)
        z = self.encoder(x)
        x_hat = self.decoder(z)
        return x_hat

    def training_step(self, batch, batch_idx):
        # training_step defines the train loop.
        # it is independent of forward
        x, y = batch
        x = x.view(x.size(0), -1)
        z = self.encoder(x)
        x_hat = self.decoder(z)
        loss = nn.functional.mse_loss(x_hat, x)
        # Logging to TensorBoard (if installed) by default
        self.log("train_loss", loss)
        return loss

    def configure_optimizers(self):
        optimizer = optim.Adam(self.parameters(), lr=1e-3)
        return optimizer


# init the autoencoder
autoencoder = LitAutoEncoder(encoder, decoder)

dataset = MNIST(os.getcwd(), download=True, transform=ToTensor())
train_loader = utils.data.DataLoader(dataset)

checkpoint_callback = ModelCheckpoint(
    dirpath=os.path.join(os.getcwd(), "checkpoints", f"run-x",),
    monitor="train_loss",
    save_top_k=3,
    mode="min",
)

trainer = pl.Trainer(
    limit_train_batches=100, max_epochs=10, callbacks=[checkpoint_callback]
)

trainer.fit(model=autoencoder, train_dataloaders=train_loader)

trainer.predict(dataloaders=train_loader, ckpt_path="last")
```

### Error messages and logs

```
GPU available: True (cuda), used: True
TPU available: False, using: 0 TPU cores
IPU available: False, using: 0 IPUs
HPU available: False, using: 0 HPUs
LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]

  | Name    | Type       | Params
---------------------------------------
0 | encoder | Sequential | 50.4 K
1 | decoder | Sequential | 51.2 K
---------------------------------------
101 K     Trainable params
0         Non-trainable params
101 K     Total params
0.407     Total estimated model params size (MB)
C:\Users\user\Anaconda3\envs\trgl\lib\site-packages\pytorch_lightning\trainer\connectors\data_connector.py:430: PossibleUserWarning: The dataloader, train_dataloader, does not have many workers which may be a bottleneck. Consider increasing the value of the `num_workers` argument` (try 12 which is the number of cpus on this machine) in the `DataLoader` init to improve performance.
  rank_zero_warn(
Epoch 9: 100%|██████████| 100/100 [00:00<00:00, 274.22it/s, v_num=5]
`Trainer.fit` stopped: `max_epochs=10` reached.
C:\Users\user\Anaconda3\envs\trgl\lib\site-packages\pytorch_lightning\trainer\connectors\checkpoint_connector.py:189: UserWarning: .predict(ckpt_path="last") is set, but there is no last checkpoint available. No checkpoint will be loaded.
  rank_zero_warn(
LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]
C:\Users\user\Anaconda3\envs\trgl\lib\site-packages\pytorch_lightning\trainer\connectors\data_connector.py:430: PossibleUserWarning: The dataloader, predict_dataloader, does not have many workers which may be a bottleneck. Consider increasing the value of the `num_workers` argument` (try 12 which is the number of cpus on this machine) in the `DataLoader` init to improve performance.
  rank_zero_warn(
Predicting DataLoader 0:  61%|██████    | 36659/60000 [00:38<00:24, 945.32it/s]C:\Users\user\Anaconda3\envs\trgl\lib\site-packages\pytorch_lightning\trainer\call.py:54: UserWarning: Detected KeyboardInterrupt, attempting graceful shutdown...
  rank_zero_warn("Detected KeyboardInterrupt, attempting graceful shutdown...")
Predicting DataLoader 0:  61%|██████    | 36659/60000 [00:38<00:24, 940.85it/s]

```

### Environment

<details>
  <summary>Current environment</summary>

* CUDA:
	- GPU:
		- Quadro RTX 3000
	- available:         True
	- version:           11.8
* Lightning:
	- lightning-utilities: 0.8.0
	- pytorch-lightning: 2.0.2
	- torch:             2.0.1
	- torchmetrics:      0.11.4
	- torchvision:       0.15.2
* Packages:
	- absl-py:           1.4.0
	- aiohttp:           3.8.4
	- aiosignal:         1.3.1
	- anyio:             3.5.0
	- appdirs:           1.4.4
	- argon2-cffi:       21.3.0
	- argon2-cffi-bindings: 21.2.0
	- asttokens:         2.0.5
	- async-timeout:     4.0.2
	- attrs:             22.1.0
	- babel:             2.11.0
	- backcall:          0.2.0
	- beautifulsoup4:    4.12.2
	- black:             23.3.0
	- bleach:            4.1.0
	- brotlipy:          0.7.0
	- cachetools:        5.3.0
	- certifi:           2022.12.7
	- cffi:              1.15.1
	- cfgv:              3.3.1
	- charset-normalizer: 2.0.4
	- click:             8.1.3
	- colorama:          0.4.6
	- comm:              0.1.2
	- contourpy:         1.0.7
	- cryptography:      39.0.1
	- cycler:            0.11.0
	- debugpy:           1.5.1
	- decorator:         5.1.1
	- defusedxml:        0.7.1
	- distlib:           0.3.6
	- docker-pycreds:    0.4.0
	- editdistance:      0.6.2
	- entrypoints:       0.4
	- exceptiongroup:    1.1.1
	- executing:         0.8.3
	- fastjsonschema:    2.16.2
	- filelock:          3.12.0
	- fonttools:         4.39.3
	- frozenlist:        1.3.3
	- fsspec:            2023.5.0
	- gitdb:             4.0.10
	- gitpython:         3.1.31
	- google-auth:       2.17.3
	- google-auth-oauthlib: 1.0.0
	- grpcio:            1.54.0
	- identify:          2.5.24
	- idna:              3.4
	- iniconfig:         2.0.0
	- ipykernel:         6.19.2
	- ipython:           8.12.0
	- ipython-genutils:  0.2.0
	- ipywidgets:        8.0.4
	- isort:             5.12.0
	- jedi:              0.18.1
	- jinja2:            3.1.2
	- joblib:            1.2.0
	- json5:             0.9.6
	- jsonschema:        4.17.3
	- jupyter:           1.0.0
	- jupyter-client:    8.1.0
	- jupyter-console:   6.6.3
	- jupyter-core:      5.3.0
	- jupyter-server:    1.23.4
	- jupyterlab:        3.5.3
	- jupyterlab-pygments: 0.1.2
	- jupyterlab-server: 2.22.0
	- jupyterlab-widgets: 3.0.5
	- kiwisolver:        1.4.4
	- lightning-utilities: 0.8.0
	- lxml:              4.9.2
	- markdown:          3.4.3
	- markupsafe:        2.1.1
	- matplotlib:        3.7.1
	- matplotlib-inline: 0.1.6
	- mistune:           0.8.4
	- mpmath:            1.2.1
	- multidict:         6.0.4
	- mypy-extensions:   1.0.0
	- nbclassic:         0.5.5
	- nbclient:          0.5.13
	- nbconvert:         6.5.4
	- nbformat:          5.7.0
	- nest-asyncio:      1.5.6
	- networkx:          2.8.4
	- nodeenv:           1.7.0
	- notebook:          6.5.4
	- notebook-shim:     0.2.2
	- numpy:             1.24.3
	- numpyencoder:      0.3.0
	- oauthlib:          3.2.2
	- packaging:         23.0
	- pandas:            2.0.1
	- pandocfilters:     1.5.0
	- parso:             0.8.3
	- pathspec:          0.11.1
	- pathtools:         0.1.2
	- patsy:             0.5.3
	- pickleshare:       0.7.5
	- pillow:            9.5.0
	- pip:               23.0.1
	- platformdirs:      3.5.0
	- pluggy:            1.0.0
	- ply:               3.11
	- pre-commit:        3.3.1
	- prometheus-client: 0.14.1
	- prompt-toolkit:    3.0.36
	- protobuf:          4.23.0
	- psutil:            5.9.0
	- pure-eval:         0.2.2
	- pyasn1:            0.5.0
	- pyasn1-modules:    0.3.0
	- pycparser:         2.21
	- pygments:          2.15.1
	- pyopenssl:         23.0.0
	- pyparsing:         3.0.9
	- pyqt5:             5.15.7
	- pyqt5-sip:         12.11.0
	- pyrsistent:        0.18.0
	- pysocks:           1.7.1
	- pytest:            7.3.1
	- python-dateutil:   2.8.2
	- pytorch-lightning: 2.0.2
	- pytz:              2022.7
	- pywin32:           305.1
	- pywinpty:          2.0.10
	- pyyaml:            6.0
	- pyzmq:             25.0.2
	- qtconsole:         5.4.2
	- qtpy:              2.2.0
	- requests:          2.29.0
	- requests-oauthlib: 1.3.1
	- rsa:               4.9
	- scipy:             1.10.1
	- seaborn:           0.12.2
	- send2trash:        1.8.0
	- sentry-sdk:        1.22.2
	- setproctitle:      1.3.2
	- setuptools:        66.0.0
	- shortuuid:         1.0.11
	- sip:               6.6.2
	- six:               1.16.0
	- smmap:             5.0.0
	- sniffio:           1.2.0
	- soupsieve:         2.4
	- stack-data:        0.2.0
	- statsmodels:       0.14.0
	- sympy:             1.11.1
	- tensorboard:       2.13.0
	- tensorboard-data-server: 0.7.0
	- terminado:         0.17.1
	- tinycss2:          1.2.1
	- tokenize-rt:       5.0.0
	- toml:              0.10.2
	- tomli:             2.0.1
	- torch:             2.0.1
	- torchmetrics:      0.11.4
	- torchvision:       0.15.2
	- tornado:           6.2
	- tqdm:              4.65.0
	- traitlets:         5.7.1
	- typing-extensions: 4.5.0
	- tzdata:            2023.3
	- urllib3:           1.26.15
	- virtualenv:        20.23.0
	- wandb:             0.15.2
	- wcwidth:           0.2.5
	- webencodings:      0.5.1
	- websocket-client:  0.58.0
	- werkzeug:          2.3.4
	- wheel:             0.38.4
	- widgetsnbextension: 4.0.5
	- win-inet-pton:     1.1.0
	- yarl:              1.9.2
* System:
	- OS:                Windows
	- architecture:
		- 64bit
		- WindowsPE
	- processor:         Intel64 Family 6 Model 165 Stepping 2, GenuineIntel
	- python:            3.10.11
	- release:           10
	- version:           10.0.19045

</details>

cc @borda @carmocca @awaelchli

**Comment by awaelchli** [comment]
Loading using `trainer.fit/validate/test/predict(..., ckpt_path="last")` is only valid when the callback was configured with `save_last=True`, but this isn't so obvious from the docs and should be clarified.
